esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -71,15 +71,88 @@ static const char* FEATURE_URL="DfuAccess";
 
 #define REMOVE_FILE_SDS_CONNECT_TIMEOUT (1000*15)  // 15 seconds
 
+const unsigned NODE_GROUP_CACHE_SIZE = 16;
+const unsigned NODE_GROUP_CACHE_MIN_DEFAULT = 30;
+
 const int DESCRIPTION_DISPLAY_LENGTH = 12;
 const unsigned MAX_VIEWKEYFILE_ROWS = 1000;
 const unsigned MAX_KEY_ROWS = 20;
 
 short days[12] = {31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
 
-void CWsDfuEx::init(IPropertyTree *cfg, const char *process, const char *service)
+NodeGroupCacheElement* CNodeGroupCache::readNodeGroup(const char* _groupName)
 {
+    Owned<IEnvironmentFactory> factory = getEnvironmentFactory();
+    Owned<IConstEnvironment> env = factory->openEnvironment();
+    if (!env)
+        throw MakeStringException(ECLWATCH_CANNOT_GET_ENV_INFO, "Failed to get environment information.");
+
+    unsigned nodeCount = 0;
+    bool replicateOutputs = false;
+    Owned<IPropertyTree> root = &env->getPTree();
+    Owned<IPropertyTreeIterator> it= root->getElements("Software/ThorCluster");
+    ForEach(*it)
+    {
+        IPropertyTree& cluster = it->query();
+        StringBuffer groupName;
+        getClusterGroupName(cluster, groupName);
+        if (groupName.length() && strieq(groupName.str(), _groupName))
+        {
+            replicateOutputs = cluster.getPropBool("@replicateOutputs", false);
+            if (replicateOutputs)
+                nodeCount = cluster.getCount("ThorSlaveProcess");
+            break;
+        }
+    }
+
+    Owned<NodeGroupCacheElement> e = new NodeGroupCacheElement(_groupName, replicateOutputs && (nodeCount > 1));
+    if (cacheSize > 0)
+    {
+        if (cache.size() >= cacheSize)
+            cache.pop_front();
+
+        cache.push_back(e.get());
+    }
+    return e.getLink();
+}
+
+NodeGroupCacheElement* CNodeGroupCache::lookup(IEspContext &context, const char* groupName, unsigned timeOutMin)
+{
+    CriticalBlock block(crit);
+
+    if (cache.size() < 1)
+        return readNodeGroup(groupName);
+
+    //erase data if it should be
+    CDateTime timeNow;
+    timeNow.setNow();
+    timeNow.adjustTime(-timeOutMin);
+    while (true)
+    {
+        std::list<Linked<NodeGroupCacheElement> >::iterator list_iter = cache.begin();
```

Comment:
Are you talking about something like: SuperHashTableOf<CNodeGroupCacheElement, const char>? I will take a look.